cranelift/frontend/src/frontend/safepoints.rs

```diff
@@ -82,350 +98,686 @@ impl<T> SlotSizeMap<T> {
-    /// Insert spills for every value that needs to be in a stack map at every
-    /// safepoint.
-    ///
-    /// We begin with a very simple, imprecise, and overapproximating liveness
-    /// analysis. This considers any use (regardless if that use produces side
-    /// effects or flows into another instruction that produces side effects!)
-    /// of a needs-stack-map value to keep the value live. This allows us to do
-    /// this liveness analysis in a single post-order traversal of the IR,
-    /// without any fixed-point loop. The result of this analysis is the mapping
-    /// from each needs-stack-map value that is live across a safepoint to its
-    /// associated stack slot.
-    ///
-    /// Finally, we spill each of the needs-stack-map values that are live
-    /// across a safepoint to their associated stack slot upon definition, and
-    /// insert reloads from that stack slot at each use of the value.
-    pub(super) fn insert_safepoint_spills(&mut self) {
+/// A set of live values.
+type LiveSet = HashSet<ir::Value>;
+
+/// A worklist of blocks' post-order indices that we need to process.
+#[derive(Default)]
+struct Worklist {
+    /// Stack of blocks to process.
+    stack: Vec<u32>,
+
+    /// The set of blocks that need to be updated. This is a subset of the
+    /// elements present in `self.stack` because `self.stack` is allowed to have
+    /// duplicates, and once we pop the first occurrence of a duplicate, we
+    /// remove it from this set, since it no longer needs updates at that
+    /// point. This potentially uses more stack space than necessary, but
+    /// prefers processing immediate predecessors, and therefore inner loop
+    /// bodies before continuing to process outer loop bodies. This ultimately
+    /// results in fewer iterations required to reach a fixed point.
+    need_updates: HashSet<u32>,
+}
+
+impl Extend<u32> for Worklist {
+    fn extend<T>(&mut self, iter: T)
+    where
+        T: IntoIterator<Item = u32>,
+    {
+        for block_index in iter {
+            self.push(block_index);
+        }
+    }
+}
+
+impl Worklist {
+    fn clear(&mut self) {
+        let Worklist {
+            stack,
+            need_updates,
+        } = self;
+        stack.clear();
+        need_updates.clear();
+    }
+
+    fn reserve(&mut self, capacity: usize) {
+        let Worklist {
+            stack,
+            need_updates,
+        } = self;
+        stack.reserve(capacity);
+        need_updates.reserve(capacity);
+    }
+
+    fn push(&mut self, block_index: u32) {
+        // Mark this block as needing an update, since we have (re?) pushed it.
+        self.need_updates.insert(block_index);
+        self.stack.push(block_index);
+    }
+
+    fn pop(&mut self) -> Option<u32> {
+        while let Some(block_index) = self.stack.pop() {
+            // If this block was pushed multiple times, we only need to update
+            // it once, so remove it from the need-updates set.
```

Comment:
This is pretty clever and it took me a second to understand why we don't simply avoid pushing duplicates; it's to do with the innermost-first heuristic mentioned above (so we want to process at the *last* duplicate pushed, not the first), is that right?

It might be worth a comment here explicitly noting that; we push multiple times because in effect we're "hoisting" the point at which we consider a block sooner into the future, and it would be too expensive to actually remove it from deeper in the stack.

(I usually write worklist algorithms with "if already in set, don't add to queue" logic and I've seen it that way more commonly so it's worth noting this trick here!)

Reply:
> This is pretty clever and it took me a second to understand why we don't simply avoid pushing duplicates; it's to do with the innermost-first heuristic mentioned above (so we want to process at the _last_ duplicate pushed, not the first), is that right?

Correct, filtering on `push` means we process the least-recently-pushed duplicate entry. Filtering on `pop` means we process the most-recently-pushed duplicate entry, which gives us that inner-loops-first behavior that is desirable for minimizing the number of iterations until we reach the fixed-point.

Added some more comments; like your "hoisting" wording.